circuits/core/src/merkle_path/mod.rs

```diff
@@ -0,0 +1,248 @@
+        let mut leaf_bits = leaf.to_bits_le();
+        leaf_bits.resize(byte_size, Boolean::new(Mode::Constant, false));
+
+        let field_input: Vec<Field<E>> = leaf_bits.chunks(num_bits).map(Field::from_bits_le).collect();
+        let mut curr_hash = self.crh.hash(&field_input);
+
+        // To traverse up a MT, we iterate over the path from bottom to top
+
+        // At any given bit, the bit being 0 indicates our currently hashed value is the left,
+        // and the bit being 1 indicates our currently hashed value is on the right.
+        // Thus `left_hash` is the sibling if bit is 1, and it's the computed hash if bit is 0
+        for (bit, sibling) in self.traversal.iter().zip_eq(&self.path) {
+            let left_hash = Field::ternary(bit, sibling, &curr_hash);
+            let right_hash = Field::ternary(bit, &curr_hash, sibling);
+
+            // TODO (raychu86) Avoid this by using the Fields as input.
+            let mut left_hash_bits = left_hash.to_bits_le();
+            let mut right_hash_bits = right_hash.to_bits_le();
+
+            // Pad the bits to the input size according to the native PoseidonCRH function.
+            let input_field_elements: Vec<Field<E>> = {
+                left_hash_bits.resize(byte_size, Boolean::new(Mode::Constant, false));
+                right_hash_bits.resize(byte_size, Boolean::new(Mode::Constant, false));
+
+                let mut input_bits: Vec<Boolean<E>> = [left_hash_bits, right_hash_bits].concat();
+                assert!(input_bits.len() <= input_size_bits, "PoseidonCRH input bits exceeds supported input size");
+
+                if input_bits.len() < input_size_bits {
+                    input_bits.resize(input_size_bits, Boolean::new(Mode::Constant, false));
+                }
+
+                input_bits.chunks(num_bits).map(Field::from_bits_le).collect()
+            };
+
+            curr_hash = self.crh.hash(&input_field_elements);
+        }
+
+        curr_hash
+    }
+
+    // TODO (raychu86): Ideally use the following implementation without bit padding.
+    // pub fn calculate_root(&self, leaf: Field<E>) -> Field<E> {
+    //     let mut curr_hash = self.crh.hash(&[leaf]);
+    //
+    //     // To traverse up a MT, we iterate over the path from bottom to top
+    //
+    //     // At any given bit, the bit being 0 indicates our currently hashed value is the left,
+    //     // and the bit being 1 indicates our currently hashed value is on the right.
+    //     // Thus `left_hash` is the sibling if bit is 1, and it's the computed hash if bit is 0
+    //     for (bit, sibling) in self.traversal.iter().zip_eq(&self.path) {
+    //         let left_hash = Field::ternary(bit, sibling, &curr_hash);
+    //         let right_hash = Field::ternary(bit, &curr_hash, sibling);
+    //
+    //         curr_hash = self.crh.hash(&[left_hash, right_hash]);
+    //     }
+    //
+    //     curr_hash
+    // }
+
+    pub fn verify(&self, leaf: Field<E>, root: Field<E>) {
```

Comment:
This should be moved to a new file called `verify.rs` with tests

Reply:
Alternative, we can remove this method. The caller can easily check that `path.to_root() == root` themselves